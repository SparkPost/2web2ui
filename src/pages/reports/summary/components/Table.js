import React, { Component } from 'react';
import { connect } from 'react-redux';
import cx from 'classnames';

import { _getTableData } from 'src/actions/summaryChart';
import typeaheadCacheSelector from 'src/selectors/reportFilterTypeaheadCache';
import { hasSubaccounts } from 'src/selectors/subaccounts';

import { TableCollection, TableHeader, Unit, Loading } from 'src/components';
import GroupByOption from './GroupByOption';
import { Empty } from 'src/components';
import { Panel } from 'src/components/matchbox';
import { GROUP_CONFIG } from './tableConfig';
import AddFilterLink from '../../components/AddFilterLink';
import _ from 'lodash';

import styles from './Table.module.scss';

export class Table extends Component {
  getColumnHeaders() {
    const { metrics, groupBy } = this.props;
    const isAggColumn = groupBy === 'aggregate';

    const primaryCol = isAggColumn
      ? null
      : {
          label: GROUP_CONFIG[groupBy].label,
          className: styles.HeaderCell,
          sortKey: GROUP_CONFIG[groupBy].keyName,
        };

    const metricCols = metrics.map(({ label, key }) => ({
      key,
      label: <div className={styles.RightAlign}>{label}</div>,
      className: cx(styles.HeaderCell, styles.NumericalHeader),
      sortKey: isAggColumn ? null : key,
    }));

    return [primaryCol, ...metricCols];
  }

  getSubaccountFilter = subaccountId => {
    const { typeaheadCache } = this.props;

    if (subaccountId === 0) {
      return { type: 'Subaccount', value: 'Primary Account (ID 0)', id: 0 };
    }

    const subaccount = _.find(typeaheadCache, { type: 'Subaccount', id: subaccountId });
    const value = _.get(subaccount, 'value') || `Deleted (ID ${subaccountId})`;
    return { type: 'Subaccount', value, id: subaccountId };
  };

  getRowData = () => {
    const { metrics, groupBy } = this.props;
    const group = GROUP_CONFIG[groupBy];

    return row => {
      const filterKey = row[group.keyName];
      const newFilter =
        group.label === 'Subaccount'
          ? this.getSubaccountFilter(filterKey)
          : { type: group.label, value: filterKey };

      const primaryCol =
        groupBy === 'aggregate' ? (
          'Aggregate Total'
        ) : (
          <AddFilterLink newFilter={newFilter} reportType="summary" content={newFilter.value} />
        );
      const metricCols = metrics.map(metric => (
        <div className={styles.RightAlign}>
          <Unit value={row[metric.key]} unit={metric.unit} />
        </div>
      ));

      return [primaryCol, ...metricCols];
    };
  };

  getDefaultSortColumn = () => {
    const { metrics } = this.props;
    return metrics[0].key;
  };

  renderAggregateTable() {
    const { tableData } = this.props;

    return (
      <TableCollection
        headerComponent={() => <TableHeader columns={this.getColumnHeaders()} />}
        getRowData={this.getRowData()}
        rows={tableData}
      />
    );
  }

  renderGroupByTable() {
    const { tableData, groupBy } = this.props;
    const rowKeyName = GROUP_CONFIG[groupBy].keyName;

    return (
      <TableCollection
        rowKeyName={rowKeyName}
        columns={this.getColumnHeaders()}
        getRowData={this.getRowData()}
        pagination
        defaultPerPage={10}
        rows={tableData}
        defaultSortColumn={this.getDefaultSortColumn()}
        defaultSortDirection="desc"
      />
    );
  }

  renderTable() {
    const { tableData, tableLoading, groupBy } = this.props;

    if (tableLoading) {
      return (
        <div className={styles.LoadingSection}>
          <div className={styles.Loading}>
            <Loading />
          </div>
        </div>
      );
    }

    if (!tableData.length) {
      return <Empty message="There is no data to display" />;
    }

    return groupBy === 'aggregate' ? this.renderAggregateTable() : this.renderGroupByTable();
  }

  render() {
    const { groupBy, hasSubaccounts, tableLoading, _getTableData } = this.props;
    return (
      <Panel.LEGACY>
        <Panel.LEGACY.Section>
          <GroupByOption
            _getTableData={_getTableData}
            groupBy={groupBy}
            hasSubaccounts={hasSubaccounts}
            tableLoading={tableLoading}
          />
        </Panel.LEGACY.Section>
        {this.renderTable()}
      </Panel.LEGACY>
    );
  }
}

const mapStateToProps = state => ({
  typeaheadCache: typeaheadCacheSelector(state),
  hasSubaccounts: hasSubaccounts(state),
  ...state.summaryChart,
});
export default connect(mapStateToProps, { _getTableData })(Table);
