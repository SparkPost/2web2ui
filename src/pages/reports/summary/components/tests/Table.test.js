import React from 'react';
import { Table } from '../Table';
import { shallow } from 'enzyme';
import _ from 'lodash';

describe('Summary Table', () => {

  let wrapper;

  const props = {
    _getTableData: jest.fn(),
    refresh: jest.fn(),
    metrics: [
      { key: 'metric_1' },
      { key: 'metric_2', unit: 'millisecond' }
    ],
    groupBy: 'domain',
    typeaheadCache: [
      { type: 'Subaccount', id: 555, value: 'sub 1 name' }
    ],
    tableData: [],
    tableLoading: false,
    hasSubaccounts: false
  };

  const data = [
    { subaccount_id: 0, metric_1: 123, metric_2: 12345 },
    { subaccount_id: 555, metric_1: 123, metric_2: 12345 },
    { subaccount_id: 1010, metric_1: 123, metric_2: 12345 }
  ];

  beforeEach(() => {
    wrapper = shallow(<Table {...props} />);
  });

  it('should render with no data', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with data', () => {
    wrapper.setProps({ tableData: data });
    expect(wrapper.find('TableCollection')).toMatchSnapshot();
  });

  it('should render loading', () => {
    wrapper.setProps({ tableLoading: true });
    expect(wrapper.find('Loading')).toHaveLength(1);
  });

  it('should render group by row data', () => {
    wrapper.setProps({ groupBy: 'subaccount' });
    const func = wrapper.instance().getRowData();
    const results = _.flatten(data.map(func));
    expect(results).toMatchSnapshot();
  });

  it('should render with aggregate data', () => {
    const aggData = [
      { metric_1: 987, metric_2: 654, metric_3: 321 }
    ];
    const metrics = [
      { key: 'metric_1' },
      { key: 'metric_2', unit: 'millisecond' },
      { key: 'metric_3' }
    ];

    wrapper.setProps({ metrics, groupBy: 'aggregate', tableData: aggData });
    expect(wrapper.find('TableCollection')).toMatchSnapshot();
  });
});
