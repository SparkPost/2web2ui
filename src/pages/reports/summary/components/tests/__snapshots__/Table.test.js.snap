// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary Table should render group by row data 1`] = `
Array [
  <Memo(Connect(AddFilterLink))
    content="Primary Account (ID 0)"
    newFilter={
      Object {
        "id": 0,
        "type": "Subaccount",
        "value": "Primary Account (ID 0)",
      }
    }
    reportType="summary"
  />,
  <div
    className="RightAlign"
  >
    <Unit
      value={123}
    />
  </div>,
  <div
    className="RightAlign"
  >
    <Unit
      unit="millisecond"
      value={12345}
    />
  </div>,
  <Memo(Connect(AddFilterLink))
    content="sub 1 name"
    newFilter={
      Object {
        "id": 555,
        "type": "Subaccount",
        "value": "sub 1 name",
      }
    }
    reportType="summary"
  />,
  <div
    className="RightAlign"
  >
    <Unit
      value={123}
    />
  </div>,
  <div
    className="RightAlign"
  >
    <Unit
      unit="millisecond"
      value={12345}
    />
  </div>,
  <Memo(Connect(AddFilterLink))
    content="Deleted (ID 1010)"
    newFilter={
      Object {
        "id": 1010,
        "type": "Subaccount",
        "value": "Deleted (ID 1010)",
      }
    }
    reportType="summary"
  />,
  <div
    className="RightAlign"
  >
    <Unit
      value={123}
    />
  </div>,
  <div
    className="RightAlign"
  >
    <Unit
      unit="millisecond"
      value={12345}
    />
  </div>,
]
`;

exports[`Summary Table should render with aggregate data 1`] = `
<TableCollection
  defaultSortColumn={null}
  defaultSortDirection="asc"
  getRowData={[Function]}
  headerComponent={[Function]}
  rows={
    Array [
      Object {
        "metric_1": 987,
        "metric_2": 654,
        "metric_3": 321,
      },
    ]
  }
/>
`;

exports[`Summary Table should render with data 1`] = `
<TableCollection
  columns={
    Array [
      Object {
        "className": "HeaderCell",
        "label": "Recipient Domain",
        "sortKey": "domain",
      },
      Object {
        "className": "HeaderCell NumericalHeader",
        "key": "metric_1",
        "label": <div
          className="RightAlign"
        />,
        "sortKey": "metric_1",
      },
      Object {
        "className": "HeaderCell NumericalHeader",
        "key": "metric_2",
        "label": <div
          className="RightAlign"
        />,
        "sortKey": "metric_2",
      },
    ]
  }
  defaultPerPage={10}
  defaultSortColumn="metric_1"
  defaultSortDirection="desc"
  getRowData={[Function]}
  pagination={true}
  rowKeyName="domain"
  rows={
    Array [
      Object {
        "metric_1": 123,
        "metric_2": 12345,
        "subaccount_id": 0,
      },
      Object {
        "metric_1": 123,
        "metric_2": 12345,
        "subaccount_id": 555,
      },
      Object {
        "metric_1": 123,
        "metric_2": 12345,
        "subaccount_id": 1010,
      },
    ]
  }
/>
`;

exports[`Summary Table should render with no data 1`] = `
<Panel.LEGACY>
  <Panel.LEGACY.Section>
    <GroupByOption
      _getTableData={[MockFunction]}
      groupBy="domain"
      hasSubaccounts={false}
      tableLoading={false}
    />
  </Panel.LEGACY.Section>
  <Empty
    message="There is no data to display"
  />
</Panel.LEGACY>
`;
